File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        if not heights:
            return 0
        stack = []
        maxArea = 0
        i = 0
        while i<= len(heights):
            h=0
            if i!= len(heights):
                h=heights[i]
            if (not stack) or h >= heights[stack[-1]]:
                stack.append(i)
                i += 1 
            else: 
                curHeight = heights[stack.pop()]
                right_min= i-1
                left_min=0
                if (stack):
                    left_min = stack[-1] + 1
                width = right_min - left_min +1
                
                maxArea = max(maxArea, (curHeight*width))
        
        return maxArea
    

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.java
class Solution {
    public static int largestRectangleArea(int[] height) {
   
    int[] less_left = new int[height.length]; 
    int[] less_right = new int[height.length];
    less_right[height.length - 1] = height.length;
    less_left[0] = -1;

    for (int i = 1; i < height.length; i++) {
        int prev = i - 1;

        while (prev >= 0 && height[prev] >= height[i]) {
            prev = less_left[prev];
        }
        less_left[i] = prev;
    }

    for (int i = height.length - 2; i >= 0; i--) {
        int next = i + 1;

        while (next < height.length && height[next] >= height[i]) {
            next = less_right[next];
        }
        less_right[i] = next;
    }

    int maxArea = 0;
    for (int i = 0; i < height.length; i++) {
        maxArea = Math.max(maxArea, height[i] * (less_right[i] - less_left[i] - 1));
    }

    return maxArea;
}
}